fix(docs-html): gen_help_html mishandles tab alignment

ychin opened the pull request:
The gen_help_html.lua script is a useful tool to generate structured HTML pages from the vim docs, but it has an issue of not properly handling tab characters under certain circumstances.

In Vim, sometimes the metadata is hidden using conceal, e.g. `*some_tag_definition*`, `|some_tag_link|`, and \`some_code_block\`. For convenience, Vim hides the '*' / '|' / '`' characters to make it more readable, but those characters still count when a tab character follows it. So if you have a tag of say 6 characters long, those two concealed characters would lead to the tab character after it starting at character 8, therefore aligned to one further tab length away. The gen_help_html doesn't account for that which is what leads to a lot of formatting errors in the generated output.

This commit fixes that by manually adding two spaces after each concealed tag, if it's followed by a tab char.

ychin commented:
Just for context I was looking into using the Neovim docs generation utility for MacVim (oh the irony) docs generation and noticed quite a few tab alignment bug. This should fix them. I haven't done much Lua scripting before so let me know if there are things to fix.

Here are some before/after screenshots:

vimindex:

<img width="268" alt="vimindex-before" src="https://user-images.githubusercontent.com/1217449/196073333-cbf238e7-5852-479c-af19-aa2b988f4d9c.png">
<img width="268" alt="vimindex-after" src="https://user-images.githubusercontent.com/1217449/196073337-6df9b60d-7169-4cda-9049-18a4a7ce54b5.png">

nvim_terminal_emulator:

<img width="542" alt="nvim_terminal_emulator-before" src="https://user-images.githubusercontent.com/1217449/196073431-bb6561bc-0402-4cb4-bedc-70a6e709b70b.png">
<img width="542" alt="nvim_terminal_emulator-after" src="https://user-images.githubusercontent.com/1217449/196073410-4b2075d3-adf7-430c-bc28-15f5d963bc45.png">

filetype:

<img width="687" alt="filetype-before" src="https://user-images.githubusercontent.com/1217449/196073501-81925e6c-7fc4-48dc-8f3a-e50dbdf0f64b.png">
<img width="687" alt="filetype-after" src="https://user-images.githubusercontent.com/1217449/196073512-8d8b2cef-ade3-4f69-964d-38712159318f.png">

ychin commented:
Also, is https://neovim.io/doc/user/ using a different CSS from the one that the script outputs?

zeertzjq commented:
There is also <https://neovim.io/doc/user2/>

ychin commented:
> There is also https://neovim.io/doc/user2/

Oh, wait, what's the difference between user and user2? Are they both generated using this tool? FWIW I can see the same alignment issue at user2 as well (e.g. [vimindex](https://neovim.io/doc/user2/vimindex.html))

zeertzjq commented:
Oh, it seems <https://neovim.io/doc/user2> is outdated and just an older version of <https://neovim.io/doc/user>

justinmk commented:
>  I was looking into using the Neovim docs generation utility for MacVim (oh the irony) docs generation and noticed quite a few tab alignment bugs.

That is very cool! PRs are welcome to make the script a bit more flexible (while staying simple and maintainable) so that it's more useful as a general purpose tool (instead of being only for Nvim core docs). Eventually I would like to put it in `runtime/`, but need to think about what that would look like. Then people could easily use this for plugins, too...

> Oh, it seems https://neovim.io/doc/user2 is outdated and just an older version of https://neovim.io/doc/user

Yeah, need to delete /user2/ from https://github.com/neovim/doc

> This commit fixes that by manually adding two spaces after each concealed tag, if it's followed by a tab char.

The results look good, we'll need to keep an eye out for regressions.

justinmk commented on scripts/gen_help_html.lua:
```diff
@@ -512,6 +532,15 @@ local function visit_node(root, level, lang_tree, headings, opt, stats)
+    local next_node_text = node_text(root:next_sibling())
+    if string.sub(next_node_text,1,1) == "\t" then
+        s = s .. '  '
+    end
```
Should be easy to extract this logic to a function (suggested name: `fix_tab_after_conceal()`), then we can avoid duplicating this everywhere.

Nit: 2-space indent :)

clason replied:
Just run `stylua scripts/gen_help_html.lua` from the repo root ;)

ychin replied:
Ok I refactored the code to the function and fixed the indent.

I didn't run stylua though because it would fix up a bunch of other parts of the code :)

clason commented on scripts/gen_help_html.lua:
```diff
@@ -369,6 +369,18 @@ local function visit_validate(root, level, lang_tree, opt, stats)
+  -- Vim tabs take into account the two concealed characters even though they
+  -- are invisible, so we need to add back in the two spaces if this is
+  -- followed by a tab to make the tab alignment to match Vim's behavior.
+  if string.sub(next_node_text,1,1) == "\t" then
```
```suggestion
  if string.sub(next_node_text,1,1) == '\t' then
```

ychin replied:
Ok fixed

ychin commented:
> The results look good, we'll need to keep an eye out for regressions.

Yeah I couldn't find a way to easily detect them other than visual inspection since the results of this have to be physically rendered out to a web page for us to see.

justinmk commented on scripts/gen_help_html.lua:
```diff
@@ -920,7 +942,7 @@ local function gen_css(fname)
 
 function M._test()
-  tagmap = get_helptags('./build/runtime/doc')
+  tagmap = './build/runtime/doc'
```
looks like a mistake

ychin replied:
Oh crap you are right. Fixed.

justinmk commented on scripts/gen_help_html.lua:
```diff
@@ -512,6 +532,8 @@ local function visit_node(root, level, lang_tree, headings, opt, stats)
     end
     local el = in_heading and 'span' or 'code'
     local s = ('%s<a name="%s"></a><%s class="%s">%s</%s>'):format(ws(), url_encode(tagname), el, cssclass, trimmed, el)
+    s = fix_tab_after_conceal(s, node_text(root:next_sibling()))
```
In all cases, this also only is needed if `opt.old` is true

ychin replied:
Yeah I was being a little lazy there. I think with or without the two-space insertion it still generates the same rendered output but then it's not necessary to add the two spaces for new-style ones then. Added the check. 

Just to confirm, after my change, now we can see that the new style docs weren't changed:

```
➜ ychin ~/Dev/external/neovim/target% diff dir/api.html dir_clean/api.html 
5064c5064
<         Generated at 2022-10-19 17:23 from <code><a href="https://github.com/neovim/neovim/commit/?">?</a></code>
---
>         Generated at 2022-10-19 17:22 from <code><a href="https://github.com/neovim/neovim/commit/?">?</a></code>
➜ ychin ~/Dev/external/neovim/target% diff dir/nvim.html dir_clean/nvim.html
169c169
<         Generated at 2022-10-19 17:23 from <code><a href="https://github.com/neovim/neovim/commit/?">?</a></code>
---
>         Generated at 2022-10-19 17:22 from <code><a href="https://github.com/neovim/neovim/commit/?">?</a></code>
```

ychin committed:
fix(docs-html): fix tab alignment for concealed characters

The gen_help_html.lua script is a useful tool to generate structured
HTML pages from the vim docs, but it has an issue of not properly
handling tab characters under certain circumstances.

In Vim, sometimes the metadata is hidden using conceal, e.g.
*some_tag_definition*, |some_tag_link|, and `some_code_block`. For
convenience, Vim hides the '*' / '|' / '`' characters to make it more
readable, but those characters still count when a tab character follows
it. So if you have a tag of say 6 characters long, those two concealed
character would lead to the tab character after it start at character 8.
The gen_help_html doesn't account for that which is what leads to a lot
of formatting errors in the generated output.

This commit fixes that by manually adding two spaces after each
concealed tag, if it's followed by a tab char.

justinmk commented:
> Also, is https://neovim.io/doc/user/ using a different CSS from the one that the script outputs?

No but it inherits the main.css from the main site (https://neovim.io/).

ychin commented:
Just a random followup, I was fixing a similar bug for https://vimhelp.org (https://github.com/c4rlo/vimhelp/pull/23) and I ended up fixing it in a more general way than this PR.

The fix I made for this PR and the original fix I had for that PR only injects two spaces if we have a 1) concealed tag/link/code, 2) followed by a hard tab immediately. It works for some cases, but it misses some other cases where we have 1) concealed tag, 2) followed by random text, 3) followed by hard tab.

I fixed this properly for vimhelp, but couldn't find an easy way to do this for Neovim's documentation, probably because my knowledge of treesitter isn't good enough. In particular, I think we need to keep an accumulated count of concealed characters and then add that many spaces right before we see a hard tab. The tricky part is to reset the concealed character count accumulation when we hit a newline (since obviously tab spacing do not persist across lines). The parser in vimhelp is line-based so it was easy to do but I didn't find an easy way to do this for Neovim's treesitter parser. Not saying it's impossible to do btw, just that it would be a little more annoying as you need to manually look for the newline and reset the accumulation.

Just in case you are wondering what I'm talking about, you can either see the examples at (https://github.com/c4rlo/vimhelp/pull/23#issuecomment-1452362775), or see this screenshot:

* https://neovim.io/doc/user/vimindex.html:

  ![image](https://user-images.githubusercontent.com/1217449/222993828-1f5f16b4-fc75-48b7-a6f4-2b565330f280.png)


* https://neovim.io/doc/user/quickref.html#Q_ma

  This one actually also contains some unrelated tree-sitter parser bugs related to the "`" character I think. This is what the text is supposed to look like:

  ```
  |m|        m{a-zA-Z}	mark current position with mark {a-zA-Z}
  |`a|       `{a-z}	go to mark {a-z} within current file
  |`A|       `{A-Z}	go to mark {A-Z} in any file
  |`0|       `{0-9}	go to the position where Vim was previously exited
  |``|       ``		go to the position before the last jump
  |`quote|   `"		go to the position when last editing this file
  |`[|       `[		go to the start of the previously operated or put text
  |`]|       `]		go to the end of the previously operated or put text
  |`<|       `<		go to the start of the (previous) Visual area
  |`>|       `>		go to the end of the (previous) Visual area
  |`.|       `.		go to the position of the last change in this file
  |'|        '{a-zA-Z0-9[]'"<>.}
              same as `, but on the first non-blank in the line
  |:marks|  :marks	print the active marks
  |CTRL-O|  N  CTRL-O	go to Nth older position in jump list
  |CTRL-I|  N  CTRL-I	go to Nth newer position in jump list
  |:ju|     :ju[mps]	print the jump list
  ```
  
  ![image](https://user-images.githubusercontent.com/1217449/222994190-db53e1e8-51f2-4779-9f70-3c471a869880.png)


---

This is a relatively rare edge case so maybe I'm the only one obsessing over the minor alignment issues, but just wanted to point them out since my PR wasn't a complete fix for the issue (it did fix most of them though).

Here're some commands if you want to find all the current problematic places (run them with pwd set to `$neovimr/runtime/doc`:

    lvimg /|\S\+|[^\t]\+\(\%9v\|\%10v\|\%17v\|\%18v\|\%25v\|\%26v\|\%33v\|\%34v\|\%41v\|\%42v\)\t/ *.txt
    lvimg /\*\S\+\*[^\t]\+\(\%9v\|\%10v\|\%17v\|\%18v\|\%25v\|\%26v\|\%33v\|\%34v\|\%41v\|\%42v\)\t/ *.txt
    lvimg /`\S\+`[^\t]\+\(\%9v\|\%10v\|\%17v\|\%18v\|\%25v\|\%26v\|\%33v\|\%34v\|\%41v\|\%42v\)\t/ *.txt

justinmk commented:
@ychin thanks for mentioning that! Are you not using gen_help_html after all?

ychin commented:
> @ychin thanks for mentioning that! Are you not using gen_help_html after all?

That's correct. I did quite like certain aspects of gen_help_html and wanted to get it to work, but there were a few issues preventing me from using it (in case you were curious of the nitty gritty details):

1. The treesitter syntax is pretty rigid. MacVim is a fork of Vim with a main [gui_mac.txt](https://github.com/macvim-dev/macvim/blob/master/runtime/doc/gui_mac.txt) as the main documentation page but there are also all sorts of changes all over the place (e.g. option.txt because we added some custom settings or in index.txt for custom commands). That means gen_help_html needs to work for all Vim documentation pages, which don't all conform to the standard "syntax". E.g. For arabic.txt you had to fix it up on Neovim side to make it work with the parser (https://github.com/neovim/neovim/commit/ef4c339fb9de87f7534303e006c281e40327f803). It does make the docs more consistent that way but I don't feel like making a giant upstream Vim PR just to reformat all the documentation for a downstream MacVim need, or maintaining permanent downstream changes that would make my life difficult every time I merge from upstream.
2. Some of the original attractions for using gen_help_html turned out not to be useful for me in the end:
   - Flow mode / new style: I was curious about flow mode but after investigating it I think gui_mac.txt would require too much changes to fit the flow mode model to work, especially in certain places that look like a table and need the pre-formatting. FWIW I'm kind of curious what you guys plan to do with it because in https://neovim.io/doc/user/api.html#api-definitions you had to make some compromises with tags like `|bufnr()|` being in a code block which prevents it from being a link. I feel like the lack of proper table formatting for flow mode makes it hard to work well with Vim-style documentation. Just shower thoughts: I wonder if retrofitting Vim documentation to fit a reflowable HTML page is inherently difficult, whereas the opposite direction of generating Vim documentation from a markup language (Markdown, LaTeX, HTML) would work better. Anyway I digress. Sometimes you can't let perfect be enemy of good.
   - I originally wanted to use the h3 headings (Vim syntax group `helpHeadline`) (which also shows up in the table of contents to the side) that gen_help_html could generate but turned out our docs don't use it too much. Also this is kind of superficial but I kind of dislike how they have to be all caps (since that's how Vim's syntax works).
       - **Bug**: As an aside I think there is a minor bug in the tree-sitter parser logic where a single character (e.g. "A" in "STARTING MACVIM FROM A TERMINAL") does not count as a headline. This is because in [grammar.json](https://github.com/neovim/tree-sitter-vimdoc/blob/master/src/grammar.json) "h3" refers to "uppercase_name", which has a regex that requires two or more characters: `[A-Z0-9.()][-A-Z0-9.()_]+`.
3. Small things:
   - gen_help_html doesn't handle `helpSpecial` tags like `[range]` (https://github.com/neovim/tree-sitter-vimdoc/issues/1).
   - Keys like `CTRL-D` do not seem to generate a link (e.g. see beginning of https://neovim.io/doc/user/index.html).

These issues are easier for Neovim to resolve because the tool is designed for Neovim where I mostly just wanted something that would not need much if any upkeep at all, and vimhelp seems more suitable as it is a simpler tool / parser that I can just run and it spits out a bunch of HTML that look correct. It's also a little easier for me if I want to change some rules in the parser to make some custom displays for MacVim's docs like slapping the MacVim logo within gui_mac.txt, but this is not the primary reason.

justinmk commented:
Thanks for the notes! 

> FWIW I'm kind of curious what you guys plan to do with it because in https://neovim.io/doc/user/api.html#api-definitions you had to make some compromises with tags like `|bufnr()|` being in a code block which prevents it from being a link. 

I don't see lack of tables as important, but there's a very feasible path to supporting it anyway: treat them as [injected markdown](https://github.com/neovim/neovim/pull/22481).

> I wonder if retrofitting Vim documentation to fit a reflowable HTML page is inherently difficult, whereas the opposite direction of generating Vim documentation from a markup language (Markdown, LaTeX, HTML) would work better. 

Yes, we're approaching from both directions. Though it's almost moot if we insist on "well-formed" vimdoc input. 

> vimhelp seems more suitable as it is a simpler tool / parser that I can just run and it spits out a bunch of HTML that look correct

I originally started with vimhelp's python regex stuff. But sure, if you're happy with the HTML that vimhelp happens to generate, that's one way to avoid... everything :)

Eventually Vim will need to fix their docs, if they care about [structural editing](https://github.com/neovim/neovim/issues/12842). Until then it's "let's pretend regex is fine".
